### PR TITLE
Standard guestbook using Cloud Datastore API rather than Objectify

### PR DESCRIPTION
**Pull request by wsh**

### CODE REVIEW

File: appengine/guestbook-cloud-datastore/README.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+Second, modify `Persistence.java`: replace `your-project-id-here` with the project ID you picked.
+
+Then start the [Cloud Datastore Emulator](https://cloud.google.com/datastore/docs/tools/datastore-emulator):
+
+    gcloud beta emulators datastore start --project=YOUR_PROJECT_ID_HERE
+
+Finally, in a new shell, [set the Datastore Emulator environmental variables](https://cloud.google.com/datastore/docs/tools/datastore-emulator#setting_environment_variables)
+and run
+
+    mvn clean appengine:devserver
```
Comment (lesv) [inline appengine/guestbook-cloud-datastore/README.md]: Looks like the old plugin, should be mvn clean appengine:run w/ new.

File: appengine/guestbook-cloud-datastore/README.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+## Deploying
+
+Modify `appengine-web.xml` to reflect your app ID and version, then:
+
+    mvn clean appengine:update
```
Comment (lesv) [inline appengine/guestbook-cloud-datastore/README.md]: mvn clean appengine:run

File: appengine/guestbook-cloud-datastore/src/main/java/com/example/guestbook/Greeting.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        Objects.equals(content, greeting.content) &&
+        Objects.equals(date, greeting.date);
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(key, authorEmail, authorId, content, date);
+  }
+}
+//[END all]
```
Comment (lesv) [inline appengine/guestbook-cloud-datastore/src/main/java/com/example/guestbook/Greeting.java]: newLine

Comment (lesv) [inline appengine/guestbook-cloud-datastore/src/main/java/com/example/guestbook/Greeting.java]: It's common to add toString() if you are doing equals & hashCode().

File: appengine/guestbook-cloud-datastore/src/main/java/com/example/guestbook/Guestbook.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+import com.google.cloud.datastore.KeyFactory;
+import com.google.cloud.datastore.Query;
+import com.google.cloud.datastore.QueryResults;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableList.Builder;
+import java.util.List;
+
+//[START all]
+public class Guestbook {
+  private static final KeyFactory kf = getKeyFactory(Guestbook.class);
```
Comment (lesv) [inline appengine/guestbook-cloud-datastore/src/main/java/com/example/guestbook/Guestbook.java]: `keys` or `keyfactory`

File: appengine/guestbook-cloud-datastore/src/main/java/com/example/guestbook/Guestbook.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+    Builder<Greeting> resultListBuilder = ImmutableList.builder();
+    while (results.hasNext()) {
+      resultListBuilder.add(new Greeting(results.next()));
+    }
+
+    return resultListBuilder.build();
+  }
+}
+//[END all]
```
Comment (lesv) [inline appengine/guestbook-cloud-datastore/src/main/java/com/example/guestbook/Guestbook.java]: newLine

File: appengine/guestbook-cloud-datastore/src/main/webapp/WEB-INF/appengine-web.xml [hunk 6]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="utf-8"?>
+<appengine-web-app xmlns="http://appengine.google.com/ns/1.0">
+    <application>your-app-id-here</application>
```
Comment (lesv) [inline appengine/guestbook-cloud-datastore/src/main/webapp/WEB-INF/appengine-web.xml]: No longer required

File: appengine/guestbook-cloud-datastore/src/main/webapp/WEB-INF/appengine-web.xml [hunk 7]
```diff
@@ -0,0 +1,10 @@
...
+    <version>your-app-version-here</version>
```
Comment (lesv) [inline appengine/guestbook-cloud-datastore/src/main/webapp/WEB-INF/appengine-web.xml]: No longer required

File: appengine/guestbook-cloud-datastore/src/main/webapp/WEB-INF/index.yaml [hunk 8]
```diff
@@ -0,0 +1,7 @@
+indexes:
+
+- kind: Greeting
+  ancestor: yes
+  properties:
+  - name: date
+    direction: desc
```
Comment (lesv) [inline appengine/guestbook-cloud-datastore/src/main/webapp/WEB-INF/index.yaml]: NewLine

File: appengine/guestbook-cloud-datastore/src/main/webapp/guestbook.jsp [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+    // view of the Greetings belonging to the selected Guestbook.
+      List<Greeting> greetings = theBook.getGreetings();
+
+    if (greetings.isEmpty()) {
+%>
+<p>Guestbook '${fn:escapeXml(guestbookName)}' has no messages.</p>
+<%
+    } else {
+%>
+<p>Messages in Guestbook '${fn:escapeXml(guestbookName)}'.</p>
```
Comment (lesv) [inline appengine/guestbook-cloud-datastore/src/main/webapp/guestbook.jsp]: YEA - I forgot this recently in some stuff -- good to see it here.

File: appengine/guestbook-cloud-datastore/README.md (hunk 2 above)
Comment (wsh) [inline appengine/guestbook-cloud-datastore/README.md]: You meant deploy, but got it.

File: appengine/guestbook-cloud-datastore/README.md (hunk 1 above)
Comment (wsh) [inline appengine/guestbook-cloud-datastore/README.md]: ack

File: appengine/guestbook-cloud-datastore/src/main/java/com/example/guestbook/Greeting.java (hunk 3 above)
Comment (wsh) [inline appengine/guestbook-cloud-datastore/src/main/java/com/example/guestbook/Greeting.java]: Fair enough, done.

File: appengine/guestbook-cloud-datastore/src/main/webapp/WEB-INF/appengine-web.xml (hunk 6 above)
Comment (wsh) [inline appengine/guestbook-cloud-datastore/src/main/webapp/WEB-INF/appengine-web.xml]: ack

File: appengine/guestbook-cloud-datastore/src/main/webapp/WEB-INF/index.yaml (hunk 8 above)
Comment (wsh) [inline appengine/guestbook-cloud-datastore/src/main/webapp/WEB-INF/index.yaml]: ack

File: appengine/guestbook-cloud-datastore/src/main/webapp/guestbook.jsp (hunk 9 above)
Comment (wsh) [inline appengine/guestbook-cloud-datastore/src/main/webapp/guestbook.jsp]: :smile: It's in the Objectify one! I just copied it.

File: appengine/guestbook-cloud-datastore/pom.xml [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+        </dependency>
+    </dependencies>
+
+    <build>
+        <!-- for hot reload of the web application-->
+        <outputDirectory>${project.build.directory}/${project.build.finalName}/WEB-INF/classes</outputDirectory>
+        <plugins>
+            <plugin>
+                <groupId>com.google.appengine</groupId>
+                <artifactId>appengine-maven-plugin</artifactId>
```
Comment (lesv) [inline appengine/guestbook-cloud-datastore/pom.xml]: Should be:
```xml
      <plugin>
        <groupId>com.google.cloud.tools</groupId>
        <artifactId>appengine-maven-plugin</artifactId>
        <version>0.1.2</version>
        <configuration>
          <!-- dev appserver configuration (standard environment only) -->
          <!--
          <devserver.host>127.0.0.1</devserver.port>
          <devserver.port>8080</devserver.port>
          -->
          <!-- staging configuration (standard/flex-compat environment only) -->
          <!--
          <stage.enableJarSplitting>true</stage.enableJarSplitting>
          -->
          <!-- staging configuration (flexible environment only) -->
          <!--
          <stage.artifact>target/some-customer-artifact.jar</stage.artifact>
          -->
          <!-- deploy configuration -->
          <deploy.promote>true</deploy.promote>
          <deploy.stopPreviousVersion>true</deploy.stopPreviousVersion>
        </configuration>
      </plugin>
```

**Comment by lesv** [comment]
@ludoch @jerjou @tswast @GoogleCloudPlatform/cloud-tools-for-java FYI - Veneer Datastore on GAE Standard

Thank you to @wsh

**Pull request merged**